packages/shell-evaluator/src/shell-evaluator.ts

```diff
@@ -14,18 +15,26 @@ type ResultHandler<EvaluationResultType> = (value: any) => EvaluationResultType
 class ShellEvaluator<EvaluationResultType = ShellResult> {
   private internalState: ShellInternalState;
   private resultHandler: ResultHandler<EvaluationResultType>;
+  private hasAppliedAsyncWriterRuntimeSupport = true;
 
   constructor(internalState: ShellInternalState, resultHandler: ResultHandler<EvaluationResultType> = toShellResult as any) {
     this.internalState = internalState;
     this.resultHandler = resultHandler;
+    if (process.env.MONGOSH_ASYNC_REWRITER2) {
+      process.emitWarning(new Error('Using @mongosh/async-rewriter2'));
+      this.internalState.asyncWriter = new AsyncWriter();
+      this.hasAppliedAsyncWriterRuntimeSupport = false;
+    }
   }
 
   public revertState(): void {
-    this.internalState.asyncWriter.symbols.revertState();
+    // eslint-disable-next-line chai-friendly/no-unused-expressions
+    (this.internalState.asyncWriter as any)?.symbols?.revertState();
   }
 
   public saveState(): void {
-    this.internalState.asyncWriter.symbols.saveState();
+    // eslint-disable-next-line chai-friendly/no-unused-expressions
+    (this.internalState.asyncWriter as any)?.symbols?.saveState();
```

Comment:
The `as any` stuff seems fine as it’s temporary until one of the async rewriters goes away. :)